python/feathub/metric_stores/prometheus_metric_store.py

```diff
@@ -0,0 +1,164 @@
+    ),
+    ConfigDef(
+        name=DELETE_ON_SHUTDOWN_CONFIG,
+        value_type=bool,
+        description=DELETE_ON_SHUTDOWN_DOC,
+        default_value=True,
+    ),
+]
+
+
+class PrometheusMetricStoreConfig(MetricStoreConfig):
+    def __init__(self, props: Dict[str, Any]) -> None:
+        super().__init__(props)
+        self.update_config_values(prometheus_metric_store_config_defs)
+
+
+class PrometheusMetricStore(MetricStore):
+    """
+    A metric store reporting metrics to Prometheus through push gateway. Metrics
+    would be reported in the following format:
+
+    - metric name: "feathub_{namespace}_{scope}_{metric_type}". For metrics defined
+                   on Features, scope is "feature".
+    - tags: the metric tags along with the following tags.
+        - feature_name: The name of the host feature
+        - table_name: The name of the sink where the host features would be written to.
+        - job: The namespace of the metric store.
+    """
+
+    def __init__(self, props: Dict) -> None:
+        """
+        :param props: The metric store properties.
+        """
+        super().__init__(props)
+        prometheus_metric_store_config = PrometheusMetricStoreConfig(props)
+        self.host_url = prometheus_metric_store_config.get(HOST_URL_CONFIG)
+        self.delete_on_shutdown = prometheus_metric_store_config.get(
+            DELETE_ON_SHUTDOWN_CONFIG
+        )
+
+    def materialize_metrics_if_any(
+        self,
+        feature_descriptor: TableDescriptor,
+        data_sink: Sink,
+    ) -> Optional[MaterializationDescriptor]:
+        if not any(
+            feature.metrics for feature in feature_descriptor.get_output_features()
+        ):
+            return None
+        return MaterializationDescriptor(
+            feature_descriptor=self._get_metrics_view(feature_descriptor, data_sink),
+            sink=self._get_metrics_sink(data_sink),
+            allow_overwrite=True,
+        )
+
+    def _get_metric_name(self, metric: Metric) -> str:
+        return f"feathub_{self.namespace}_feature_{metric.metric_type}"
+
+    def _get_metric_tags(
+        self, metric: Metric, feature: Feature, data_sink: Sink
+    ) -> Dict[str, str]:
+        return {
+            **metric.get_tags(),
+            "feature_name": feature.name,
+            "table_name": data_sink.name,
+        }
+
+    # TODO: support treating zero window as infinite window in SlidingWindowTransform
+    def _get_metrics_view(
+        self, features_desc: TableDescriptor, data_sink: Sink
+    ) -> TableDescriptor:
+        metric_features = []
+        post_process_metric_features = []
+        for feature in features_desc.get_output_features():
+            for metric in feature.metrics:
+                metric_tags = self._get_metric_tags(metric, feature, data_sink)
+                metric_tags["metric_name"] = self._get_metric_name(metric)
+                metric_feature_name = generate_random_name("feathub_metric")
```

Comment:
Would it be simpler to initialize the metric_feature_name according to the metric name pattern specified in `PrometheusMetricStore`'s python doc?

Reply:
Given that each metric need to generate a sequence of transformations and we need columns to store the intermediate result, and that metrics might have identical metric names, the practice of generating random metric name is still kept in the latest implementation.